src/steamship/agents/agents.py

```diff
@@ -0,0 +1,187 @@
+            Block(
+                text="Your task has been scheduled.",
+                tags=[Tag(kind="Task ID", value={"task_id", task.task_id})],
+            )
+        ]
+
+    def _schedule_task(self, time: str, task_kwargs: dict) -> Task:
+        url = self.client._url(operation="package/instance/invoke")
+        payload = {
+            "instanceHandle": self._handle,
+            "payload": {
+                "httpVerb": "POST",
+                "invocationPath": self._method,
+                "arguments": task_kwargs,
+            },
+        }
+        return self._call(url, payload, time)
+
+    def _call(self, url, payload, schedule_time) -> Task:
+        # does the steamship-appropriate calling function.
+        pass
+
+
+class ExampleCacheTool(WorkspaceTool):
+    name = "CacheTool"
+    human_description = "In-Memory Key-Value Store"
+    ai_description = ("Used to store and lookup values by a known key.",)
+    store: KeyValueStore
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.store = KeyValueStore(client=self.client, store_identifier=f"cache-{uuid.uuid4()}")
+
+    def run(self, tool_input: List[Block], context: AgentContext) -> List[Block]:
+        pass
+
+
+class Agent(BaseModel, ABC):
+    tools: Optional[Tool] = []
+
+    @abstractmethod
+    def next_action(self, context: AgentContext) -> Action:
+        pass
+
+    @abstractmethod
+    def execute_action(self, action: Action, context: AgentContext):
+        pass
+
+    @abstractmethod
+    def run(self, agent_input: List[Block], context: AgentContext) -> List[Block]:
+        pass
+
+
+REACT_PROMPT = "<insert prompt that works here>."
+
+
+class ReACTAgent(Agent):
+    client: Steamship
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._llm = self.client.use_plugin("gpt-4")
+
+    def _extract_action(self, blocks: [Block]) -> Action:
+        # whatever format matches the prompt...
+        pass
+
+    def next_action(self, context: AgentContext) -> Action:
+        # assuming most recent block is stored in context ??
+        action_task = self._llm.generate(...)
+        action_task.wait()
+        return self._extract_action(action_task.output.blocks)
+
+    def execute_action(self, action: Action, context: AgentContext):
+        tool_name = action.name
+        tool = self.tools.get(tool_name)
+        # todo: how to track new blocks? in context? somewhere else?
+        new_blocks = tool.run(action.data, context)
+        context.update_blocks(new_blocks)
+
+    def run(self, agent_input: List[Block], context: AgentContext) -> List[Block]:
+        # push agent_input somewhere?  into context?
+        while not isinstance(action := self.next_action(context), FinishAction):
+            self.execute_action(action=action, context=context)
+
+
+class ChatHistory(BaseModel):
+    pass
+
+
+class ChatReACTAgent(ReACTAgent):
+    client: Steamship
+    chat_history: ChatHistory
```

Comment:
Is this better represented in `AgentContext`?

Reply:
This feels like a weak argument for "yes", but I do know that in the Engine, wrapping things into a Context object made it miles easier to then add new things to that context when we realized they were important.

Also I suppose it provides slightly stronger typing than relying on a kwargs list that might change over time. 

The fact that frameworks like Langchain so heavily prefer kwarg-plumbing makes me want to understand the thinking there more, though, too.. I guess it provides for easier ad-hoc overriding?